#!/usr/bin/env python
import glob
import os
from snakemake.utils import min_version

##include rules (can be commented out)

include: "rules/filter_vcf.smk"
include: "rules/vcfconsensus.smk"
include: "rules/find_stopcodons.smk"

#Snakemake config
min_version("5.5")
configfile: "config.yaml"

chr_list = list(range(1,23))+["X", "Y"]
chr= ["chr" + str(i) for i in chr_list]

rule all:
    input:
        expand("filtered_vcf/{chromosome}_filtered.vcf", chromosome = chr),
        "filtered_vcf/common_annotated.vcf.gz",
        "output/CHM13_new.fasta"

rule fix_header:
    input:
        vcf = config["vcf"],
    output:
        "filtered_vcf/all_chr_fixed_header.vcf"
    container:
        "docker://broadinstitute/gatk"
    threads: 1
    resources:
        nodes = 1
    shell:
        """
        gatk FixVcfHeader \
          I={input} \
          O={output}
        """

rule split_by_chr:
    input:
        vcf = 'filtered_vcf/all_chr_fixed_header.vcf',
    output:
        "filtered_vcf/{chromosome}.vcf"
    params:
        chr = "{chromosome}"
    container:
        "docker://broadinstitute/gatk"
    threads: 1
    resources:
        nodes = 1
    shell:
        """
        gatk CreateSequenceDictionary -R {config[ref_genome]}
        gatk SelectVariants \
          -R {config[ref_genome]} \
          -V {input.vcf} \
          -L {params.chr} \
          --exclude-filtered true \
          -O {output}
        """

rule mergevcf:
    input:
        expand("filtered_vcf/{chromosome}_filtered.vcf.gz", chromosome=chr)
    output:
        "filtered_vcf/common.vcf.gz"
    conda:
        "envs/utils.yaml"
    threads: 10
    resources:
        nodes = 1
    shell:
        """
        bcftools concat {input} \
                 --threads {threads} \
                 -O z \
                 -o {output} && tabix -p vcf {output}       
        """
rule annotatevcf:
    input:
        vcf = "filtered_vcf/common.vcf.gz"
        vcf_clinvar = {config[vcf_clinvar]}
    output:
        "filtered_vcf/common_annotated.vcf.gz"
    container:
        "docker://broadinstitute/gatk"
    threads: 10
    resources:
        nodes = 1
    shell:
        """
        gatk VariantAnnotator \
          -R {config[ref_genome]} \
          -V {input.vcf} \
          -O {output} \
          -resource:clinvar {input.vcf_clinvar} -E clinvar.CLNSIG
        """

